### ISSUE TITLE
Qubes-style forced window decoration

### ISSUE THREAD

**Issue by lvh** [mviereck/x11docker#91]
In Qubes, you run related software (e.g. personal, work...) in related Xen domUs for separation. In dom0, a window manager makes them visually distinct via a combination of window title prefix (which can not be set by the child VM/domU) and window decoration color. This still looks "seamless"

How do I accomplish something similar in x11docker? Can we get that with pure Wayland?

I can think of a few things that ostensibly work. `--xpra`   and `--xpra-wayland` work really well.  Window decoration appears to be mutter (I'm running Gnome 3)'s problem. Not sure if there's an obvious way to tweak window decoration color, which I've found very helpful for quickly determining what context a window lives in.

**Comment by lvh** [comment]
Secondly, of course, we can run Weston or whatever, but now we're purely relying on client side decorations, which is great as long as the client isn't trying to trick you with a UI redress attack.

**Comment by mviereck** [comment]
I have added a new option `--border` to draw a border into xpra windows. It follows the xpra syntax desribed above. A color argument can be specified optionally, otherwise it defaults to a decent `--border=blue,1`.

**Comment by lvh** [comment]
Whoa, that's some maintainer response! Thanks :)

Re: xpra; you're right! I was working with `jess/cathode` specifically and that seems to, somehow, set the hostname to the parent hostname -- but that seems fine.

When using  `--xpra-wayland` I get a native-looking Wayland app (I get that XWayland is in the center). It looks native in my Wayland compositor (sway) -- and IIUC, that means the window decoration (e.g. title) is not controlled by the container?

**Comment by mviereck** [comment]
> I was working with jess/cathode specifically and that seems to, somehow, set the hostname to the parent hostname -- but that seems fine.

You are right - xpra shows the host-hostname with `jess/cathode`, but the container-hostname with others. I am a bit confused why.

> that means the window decoration (e.g. title) is not controlled by the container?

Yes. The container application sets a window title, but xpra controls it. 

x11docker runs two instances of xpra, *xpra server* and *xpra client*. xpra is designed to run across a network where these two instances would run on different machines.

The container application runs on an invisible X server: Xvfb, Xdummy or Xwayland, a virtual frame buffer (vfb) in xpra terms. 

On that vfb *xpra server* acts as a window manager. *xpra server* forwards the window content from the vfb to *xpra client*. The host X server has its own window manager where *xpra client* is a regular X client.

Because *xpra client* is the real client of the host X server, it controls the window title. It can be customized. From `man xpra`:
```
 --title=VALUE
              Sets  the  text shown as window title.  The string supplied can make use of remote meta‐
              data placeholders which will be populated at runtime with the  values  from  the  remote
              server.  The default value used is "@title@ on @client-machine@".

              The following placeholders are defined:

              @title@
                     Will be replaced by the remote window's title.

              @client-machine@
                     Will be replaced by the remote server's hostname.
```

> When using --xpra-wayland I get a native-looking Wayland app (I get that XWayland is in the center). It looks native in my Wayland compositor (sway) 

If you run x11docker with `--xpra` or `--xpra-xwayland` on sway, *xpra client* will run in both cases as an X client of sway. sway runs Xwayland to support X applications.
The difference between `--xpra` and `--xpra-xwayland` is only a different vfb. This difference is present on *xpra server* side. Xwayland from `--xpra-xwayland` has nothing to do with Xwayland from sway.
In any case, it does not run as a Wayland client of sway.

Is this comprensible, or did I confuse you completly? :-p

**Comment by lvh** [comment]
(Sorry, I haven't had time to work on this, but I hope to pick up up Soon(TM). Everything you said made sense.)

**Comment by mviereck** [comment]
No problem ;-). Feel free to reopen if you have more questions or wishes.
